### PR TITLE
clientv3/concurrency: Mutex concurrency

### PR DESCRIPTION
**Pull request by afking**
Adds a semaphore for concurrent access by local goroutines. Now follows behaviour of sync.Mutex more closely.

Fixes https://github.com/etcd-io/etcd/issues/11360

Removes race conditions with multiple goroutines for a single concurrent.Mutex. Previous behaviour with multiple sessions remains the same.

### CODE REVIEW

**Comment by afking** [comment]
Could Travis CI be retried? Failed on 
```
--- FAIL: TestBalancerUnderNetworkPartitionWatchLeader (3.30s)
network_partition_test.go:266: took too long to detect leader lost
```

**Comment by jingyih** [comment]
> Could Travis CI be retried? Failed on
> 
> ```
> --- FAIL: TestBalancerUnderNetworkPartitionWatchLeader (3.30s)
> network_partition_test.go:266: took too long to detect leader lost
> ```

This is unrelated. Don't worry about it.

**Comment by jingyih** [comment]
I think we need better documentation on the APIs for concurrecy.Mutex. It is unclear to me what the exact semantics are. For example, I think in current implementation, the waiters are queued and served in FIFO order. With this PR, local waiter are blocked by a local semaphore, hence losing such guarantee. Please correct me if I misread.

**Comment by afking** [comment]
@jingyih thats correct. So the FIFO behaviour is preserved for each separate sync.Mutex, like before. If you use one you get a local fifo before etcd. This might be able to be made consistent but would need to generate the mutex state for each Lock() call.

**Comment by jingyih** [comment]
> @jingyih thats correct. So the FIFO behaviour is preserved for each separate sync.Mutex, like before. If you use one you get a local fifo before etcd. This might be able to be made consistent but would need to generate the mutex state for each Lock() call.

Thanks for confirming.

After a bit more reading, I am not sure if Lock() should block in your case. In the current implementation, for example:

```
m1 := concurrency.NewMutex(s, "foo")
m2 := concurrency.NewMutex(s, "foo")
m1.Lock()
m2.Lock() // this does not block
```

So my understanding of the mutex semantic is that, a mutex is created based on a session and an identifying name string, it can be locked / unlocked by a session (as opposed to a goroutine).

**Comment by jingyih** [comment]
If we want to keep the same semantics, we should make Lock() thread safe (or mention that it is not thread safe), but it should not block in the following case:
```
m := concurrency.NewMutex(s, "foo")
m.Lock()
m.Lock() // this should not block, because m is already hold by s
```

**Comment by afking** [comment]
The two sessions should block each other as shown in the examples for [Lock()](https://godoc.org/github.com/coreos/etcd/clientv3/concurrency#Mutex.Lock) and [TryLock()](https://godoc.org/github.com/coreos/etcd/clientv3/concurrency#Mutex.TryLock). They can be locked and unlocked per session but this PR makes each session thread safe too.

As part of the API is [NewLocker](https://godoc.org/github.com/coreos/etcd/clientv3/concurrency#NewLocker) which returns a sync.Locker. I think this should be safe for multiple go routines otherwise it's very easy for users to misuse.

**Comment by afking** [comment]
Sorry I just realised in your examples the two locks on one session. I think this behaviour should be changed to be closer to sync.Mutex and block if lock is called twice. Maybe I am misunderstanding the use case of sessions?

**Comment by jingyih** [comment]
I think we can first make this function thread safe but not blocking, which is consistent with the current API behavior.

If we want to change API behavior, it can be another PR.

What do you think?

**Comment by afking** [comment]
@jingyih thanks for taking the time to review. I pushed up a change so the Lock now only protects the calls to TryLock, Lock & Unlock. This keeps the previous behaviour.

I'm a bit unsure if this is a useful addition now? Before it would have been easier to wrap concurrency.Mutex with a semaphore to get the sync.Mutex behaviour I was trying to introduce. Now theres another stage of locking which is probably not needed if all clients are synchronous. Maybe we could just make the documentation clearer?

**Comment by jingyih** [comment]
> I'm a bit unsure if this is a useful addition now? Before it would have been easier to wrap concurrency.Mutex with a semaphore to get the sync.Mutex behaviour I was trying to introduce. Now theres another stage of locking which is probably not needed if all clients are synchronous. Maybe we could just make the documentation clearer?

Do you mind elaborate your use case? (maybe with an example?)

File: clientv3/concurrency/mutex.go [hunk 1]
```diff
@@ -55,19 +64,22 @@ func (m *Mutex) TryLock(ctx context.Context) error {
 		m.hdr = resp.Header
 		return nil
 	}
-	client := m.s.Client()
-	// Cannot lock, so delete the key
-	if _, err := client.Delete(ctx, m.myKey); err != nil {
+
+	// Cannot lock so unlock
+	if err := m.unlock(ctx); err != nil {
```
Comment (xiang90) [inline clientv3/concurrency/mutex.go]: this seems strange... why do we unlock when lock fails? probably just keep the old cleanup logic. (i understand the effect the the same, but code repetition here does improve readability)

**Pull request closed without merging**